### ISSUE TITLE
ccrewrite: Rewritten assembly gets runtime error requesting rewrite

### ISSUE THREAD

**Issue by yaakov-h** [Microsoft/CodeContracts#321]
I found a very weird corner case in my code this morning.

Using the following code:

```
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;

namespace ContractGeneratorMethod
{
	class Program
	{
		public static void Main(string[] args)
		{
			Console.WriteLine("A few of my least favourite things:");

			foreach (var type in Things.LeastFavourite.AFewOfMyLeastFavouriteThings)
			{
				Console.WriteLine("  - {0}", type);
			}
		}
	}

	public static class Things
	{
		public static class LeastFavourite
		{
			// With credits to The Chaser's War on Everything
			public static IEnumerable<string> AFewOfMyLeastFavouriteThings
			{
				get
				{
					Contract.Ensures(Contract.Result<IEnumerable<string>>() != null);

					yield return "Prim Julie Andrews";
					yield return "Cute Singing Kiddies";
				}
			}
		}
	}
}
```

Here's the configurations I've tested:

Configuration | Runtime Contract Checking Enabled | The code works
---|---|---
Debug | No | :white_check_mark: 
Release | No | :white_check_mark: 
Debug | Yes | :white_check_mark: 
Release | Yes | :negative_squared_cross_mark: 

In the non-working configuration (Release with Runtime Contract Checking enabled), the code crashes with the following error:

```
[Window Title]
AAAA.vshost.exe - Postcondition failed.

[Main Instruction]
AAAA.vshost.exe - Postcondition failed.

[Content]
Description: An assembly (probably "AAAA") must be rewritten using the code contracts binary rewriter (CCRewrite) because it is calling Contract.Ensures and the CONTRACTS_FULL symbol is defined.  Remove any explicit definitions of the CONTRACTS_FULL symbol from your project and rebuild.  CCRewrite can be downloaded from http://go.microsoft.com/fwlink/?LinkID=169180. 
After the rewriter is installed, it can be enabled in Visual Studio from the project's Properties page on the Code Contracts pane.  Ensure that "Perform Runtime Contract Checking" is enabled, which will define CONTRACTS_FULL.

[Expanded Information]
Stack Trace:
   at System.Diagnostics.Contracts.Contract.Ensures(Boolean condition)
   at ContractGeneratorMethod.Things.LeastFavourite.<get_AFewOfMyLeastFavouriteThings>d__1.MoveNext() in C:\Temp\ContractGeneratorMethod\AAAA\Program.cs:line 31
   at ContractGeneratorMethod.Program.Main(String[] args) in C:\Temp\ContractGeneratorMethod\AAAA\Program.cs:line 13
   at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
   at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

[^] Hide details  [Abort] [Debug] [Ignore]
```

Looking at the IL (from VS2015 Update 1), the call to `Contract.Ensures` made it's way into the generator's `MoveNext` method. The IL for `ContractGeneratorMethod.Things.LeastFavourite.<get_AFewOfMyLeastFavouriteThings>d__1.MoveNext()` is:

```
.method private final hidebysig newslot virtual 
	instance bool MoveNext () cil managed 
{
	.override method instance bool [mscorlib]System.Collections.IEnumerator::MoveNext()
	// Method begins at RVA 0x224c
	// Code size 103 (0x67)
	.maxstack 2
	.locals init (
		[0] int32
	)

	IL_0000: ldarg.0
	IL_0001: ldfld int32 ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>1__state'
	IL_0006: stloc.0
	IL_0007: ldloc.0
	IL_0008: switch (IL_001b, IL_0043, IL_005e)

	IL_0019: ldc.i4.0
	IL_001a: ret

	IL_001b: ldarg.0
	IL_001c: ldc.i4.m1
	IL_001d: stfld int32 ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>1__state'
	IL_0022: call !!0 [mscorlib]System.Diagnostics.Contracts.Contract::Result<class [mscorlib]System.Collections.Generic.IEnumerable`1<string>>()
	IL_0027: ldnull
	IL_0028: cgt.un
	IL_002a: call void [mscorlib]System.Diagnostics.Contracts.Contract::Ensures(bool)
	IL_002f: ldarg.0
	IL_0030: ldstr "Prim Julie Andrews"
	IL_0035: stfld string ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>2__current'
	IL_003a: ldarg.0
	IL_003b: ldc.i4.1
	IL_003c: stfld int32 ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>1__state'
	IL_0041: ldc.i4.1
	IL_0042: ret

	IL_0043: ldarg.0
	IL_0044: ldc.i4.m1
	IL_0045: stfld int32 ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>1__state'
	IL_004a: ldarg.0
	IL_004b: ldstr "Cute Singing Kiddies"
	IL_0050: stfld string ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>2__current'
	IL_0055: ldarg.0
	IL_0056: ldc.i4.2
	IL_0057: stfld int32 ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>1__state'
	IL_005c: ldc.i4.1
	IL_005d: ret

	IL_005e: ldarg.0
	IL_005f: ldc.i4.m1
	IL_0060: stfld int32 ContractGeneratorMethod.Things/LeastFavourite/'<get_AFewOfMyLeastFavouriteThings>d__1'::'<>1__state'
	IL_0065: ldc.i4.0
	IL_0066: ret
} // end of method '<get_AFewOfMyLeastFavouriteThings>d__1'::MoveNext
```

This happens in both Full and ReleaseRequires configurations, and with or without Public Surface Contracts enabled.

This does **not** happen if I remove the inner class `LeastFavourite` (i.e. the property is `Things.AFewOfMyLeastFavouriteThings`. Introducing this inner class seems to break the rewritten assembly.

**Comment by hubuk** [comment]
@yaakov-h I am not able to reproduce the problem. Can you share full project? It seems that ccrewrite is not being invoked on the output assembly.

**Comment by yaakov-h** [comment]
@hubuk it's definitely being invoked, because other Contract calls such as `Contract.Requires<TException>` are being appropriately substituted with `__ContractsRuntime.Require<TException>`.

I've uploaded a full repro project [here](https://wisetechglobal-my.sharepoint.com/personal/yaakov_smith_wtg_zone/_layouts/15/guestaccess.aspx?guestaccesstoken=E2dN6nscwqScY2YV7lSaJ1DmiVSIIWBZGfgig4lMrm8%3d&docid=07aea4695c5184360904a9868e7462f78).

**Comment by hubuk** [comment]
I have already fixed the root cause. Trying to create a failing test.

**Comment by hubuk** [comment]
I did quite a mess here. Sorry for that.
Fix is available at #337. It required slightly modified test environment and resulted in a new issue for fixing ccrewrite when using oob contracts - #336.
